src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java

```diff
@@ -297,76 +297,140 @@ Set<String> verifyInputColumns(
     this.flushLock.lock();
     try {
       this.channelState.updateInsertStats(System.currentTimeMillis(), this.bufferedRowCount);
       if (onErrorOption == OpenChannelRequest.OnErrorOption.CONTINUE) {
-        // Used to map incoming row(nth row) to InsertError(for nth row) in response
-        int rowIndex = 0;
-        for (Map<String, Object> row : rows) {
-          InsertValidationResponse.InsertError error =
-              new InsertValidationResponse.InsertError(row, rowIndex);
-          try {
-            Set<String> inputColumnNames = verifyInputColumns(row, error, rowIndex);
-            rowsSizeInBytes +=
-                addRow(row, this.bufferedRowCount, this.statsMap, inputColumnNames, rowIndex);
-            this.bufferedRowCount++;
-          } catch (SFException e) {
-            error.setException(e);
-            response.addError(error);
-          } catch (Throwable e) {
-            logger.logWarn("Unexpected error happens during insertRows: {}", e);
-            error.setException(new SFException(e, ErrorCode.INTERNAL_ERROR, e.getMessage()));
-            response.addError(error);
-          }
-          checkBatchSizeEnforcedMaximum(rowsSizeInBytes);
-          rowIndex++;
-          if (this.bufferedRowCount == Integer.MAX_VALUE) {
-            throw new SFException(ErrorCode.INTERNAL_ERROR, "Row count reaches MAX value");
-          }
-        }
-        checkBatchSizeRecommendedMaximum(rowsSizeInBytes);
-      } else {
-        // If the on_error option is ABORT, simply throw the first exception
-        float tempRowsSizeInBytes = 0F;
-        int tempRowCount = 0;
-        for (Map<String, Object> row : rows) {
-          Set<String> inputColumnNames = verifyInputColumns(row, null, tempRowCount);
-          tempRowsSizeInBytes +=
-              addTempRow(row, tempRowCount, this.tempStatsMap, inputColumnNames, tempRowCount);
-          checkBatchSizeEnforcedMaximum(tempRowsSizeInBytes);
-          tempRowCount++;
-        }
-        checkBatchSizeRecommendedMaximum(tempRowsSizeInBytes);
-
-        moveTempRowsToActualBuffer(tempRowCount);
-
-        rowsSizeInBytes = tempRowsSizeInBytes;
-        if ((long) this.bufferedRowCount + tempRowCount >= Integer.MAX_VALUE) {
-          throw new SFException(ErrorCode.INTERNAL_ERROR, "Row count reaches MAX value");
-        }
-        this.bufferedRowCount += tempRowCount;
-        this.statsMap.forEach(
-            (colName, stats) ->
-                this.statsMap.put(
-                    colName,
-                    RowBufferStats.getCombinedStats(stats, this.tempStatsMap.get(colName))));
+        response = insertRowsHelperForContinue(rows, offsetToken);
+      } else if (onErrorOption == OpenChannelRequest.OnErrorOption.ABORT) {
+        response = insertRowsHelperForAbort(rows, offsetToken);
+      } else if (onErrorOption == OpenChannelRequest.OnErrorOption.SKIP_BATCH) {
+        response = insertRowsHelperForSkipBatch(rows, offsetToken);
```

Comment:
This is a good place to implement Strategy design and factory pattern. 

Being a little pedantic but this is how it would look like. (You probably would know it but just hashing it out here so there is less confusion)


```
C'tor of AbstractRowBuffer would fetch the implementation based on factory class

public class InsertionStrategyFactory {
    public static InsertionStrategy createStrategy(OnErrorOption type) {
        switch (type) {
            case CONTINUE:
                return new ContinueInsertionStrategy();
            case ABORT:
                return new AbortInsertionStrategy();
            case SKIP_BATCH:
                return new SkipBatchInsertionStrategy();
            default:
                throw new IllegalArgumentException("Unknown on error option: " + type);
        }
    }
}


public interface InsertionStrategy {
    Response insertRows(List<Row> rows, OffsetToken offsetToken);
}
// 
public class ContinueInsertionStrategy implements InsertionStrategy {

    @Override
    public Response insertRows(List<Row> rows, OffsetToken offsetToken) {
        channelState.updateInsertStats(System.currentTimeMillis(), bufferedRowCount);
        return insertRowsHelperForContinue(rows, offsetToken);
    }
}

```


Benefits:
- we can make this code more testable. 
- Abstract out different implementations. 
- In future we can add more strategies easily. 



WDYT?

Reply:
Might as well use abstract class if you think there is some common implementation in all three subclasses.